compiler/front/commands.nim

```diff
@@ -1412,6 +183,186 @@ proc processArgument*(pass: TCmdLinePass; p: OptParser;
+    ## list of cli only flags
+    cliFlagVersion
+    cliFlagHelp
+    cliFlagHelpFull
+    cliFlagHelpAdvanced
+    cliFlagMsgFormat
+
+type
+  CliEventKind* = enum
+    # errors - cli command
+    cliEvtErrInvalidCommand # main.nim
+    cliEvtErrCmdMissing # cmdlinehelper.nim
+    cliEvtErrCmdExpectedNoAdditionalArgs # nim.nim
+      ## flag disallows additional args
+    cliEvtErrFlagArgExpectedFromList # commands.nim
+      ## flag expected arg from an allow/valid list, but got none
+    cliEvtErrFlagArgNotFromValidList # commands.nim
+      ## flag expects arg from an allow/valid list
+    cliEvtErrRunCmdFailed  # commands.nim and nim.nim
+    cliEvtErrGenDependFailed # main.nim
+    # errors - general flag/option/switches (TODO: standardize on "flag")
+    cliEvtErrUnexpectedRunOpt # commands.nim and nim.nim
+    cliEvtErrFlagArgForbidden # help/version/fullhelp/etc disallow args
+    cliEvtErrFlagProcessing # commands.nim
+    # errors - specific flag/option/switch
+    cliEvtErrNoCliParamsProvided # commands.nim and nim.nim
+    # warnings - general flags/options/switches
+    cliEvtWarnSwitchValDeprecatedNoop
+    # hints - general flag/options/switches and/or processing
+    cliEvtHintPathAdded
+      ## currently only triggered if nimble adds a path
+      # xxx: this doesn't feel like a hint, more like "info" or "trace"
+
+  CliEvent* = object
+    # xxx: should these be 'flat' log events? put another way is the 
+    #      `procResult` field a design smell?
+    srcCodeOrigin*: InstantiationInfo
+    pass*: TCmdLinePass
+    case kind*: CliEventKind
+      of cliEvtErrInvalidCommand,
+          cliEvtErrCmdExpectedNoAdditionalArgs,
+          cliEvtErrUnexpectedRunOpt:
+        cmd*: string
+        unexpectedArgs*: string
+      of cliEvtErrFlagArgForbidden,
+          cliEvtErrFlagArgExpectedFromList,
+          cliEvtErrFlagArgNotFromValidList:
+        flag*: CliFlagKind
+        givenFlg*: string
+        givenArg*: string
+      of cliEvtErrRunCmdFailed,
+          cliEvtErrGenDependFailed:
+        shellCmd*: string
+        exitCode*: int
+      of cliEvtErrFlagProcessing,
+          cliEvtWarnSwitchValDeprecatedNoop:
+        origParseOptKey*, origParseOptVal*: string
+        procResult*: ProcSwitchResult
+      of cliEvtHintPathAdded:
+        pathAdded*: string
+      of cliEvtErrNoCliParamsProvided,
+          cliEvtErrCmdMissing:
+        discard
+
+const
+  cliLogAllKinds = {low(CliEventKind) .. high(CliEventKind)}
+  cliEvtErrors   = {cliEvtErrInvalidCommand .. cliEvtErrNoCliParamsProvided}
+  cliEvtWarnings = {cliEvtWarnSwitchValDeprecatedNoop}
+  cliEvtHints    = {cliEvtHintPathAdded}
+
+static:
+  const unaccountedForEvtKinds =
+    cliLogAllKinds - cliEvtErrors - cliEvtWarnings - cliEvtHints
+  doAssert unaccountedForEvtKinds == {}, "Uncategorized event kinds: " &
+                                            $unaccountedForEvtKinds
+
+proc procSwitchResultToEvents*(conf: ConfigRef, pass: TCmdLinePass,
```

Comment:
Given its current callsites, I think it might make sense to turn this into an iterator.

Reply:
Imma do this

Reply:
Done.